### ISSUE TITLE
Internal compiler error raised for `rs-graph`

### ISSUE THREAD

**Issue by lyro** [rust-lang-nursery/rust-clippy#2196]
When checking the latest version of the `rs-graph` crate (download source code from [here](https://www.mathematik.uni-kassel.de/%7Efifr/fossils/rs-graph/tarball/rs-graph-trunk.tar.gz?name=rs-graph)) with clippy, the "internal compiler error" below is raised. Normal compilation works fine without any error. I use the following versions:
```
% rustc --version
rustc 1.23.0-nightly (8b22e70b2 2017-10-31)
% cargo --version
cargo 0.24.0-nightly (859c2305b 2017-10-29)
% cargo clippy --version                                                                                                                        
0.0.168
```

```
% cargo clippy
   Compiling rs-graph v0.7.0 (file:///home/rust/rs-graph)                                                                    
error: internal compiler error: /checkout/src/librustc_typeck/collect.rs:1084: compute_type_of_item: unexpected item type: ItemTrait(Normal, Generics { lifetimes: [], ty_params: [], where_clause: WhereClause { id: NodeId(447), predicates: [BoundPredicate(WhereBoundPredicate { span: src/builder.rs:33:11: 33:22, bound_lifetimes: [], bounded_ty: type(Self), bounds: [TraitTyParamBound(PolyTraitRef { bound_lifetimes: [], trait_ref: TraitRef { path: path(Sized), ref_id: NodeId(449) }, span: src/builder.rs:33:17: 33:22 }, None)] })] }, span: src/lib.rs:1:1: 1:1 }, [], [TraitItemRef { id: TraitItemId { node_id: NodeId(450) }, name: Graph(162), kind: Type, span: src/builder.rs:36:5: 36:16, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(452) }, name: Node(160), kind: Type, span: src/builder.rs:39:5: 39:21, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(455) }, name: Edge(161), kind: Type, span: src/builder.rs:42:5: 42:21, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(458) }, name: new(510), kind: Method { has_self: false }, span: src/builder.rs:45:5: 47:6, defaultness: Default { has_value: true } }, TraitItemRef { id: TraitItemId { node_id: NodeId(466) }, name: with_capacities(511), kind: Method { has_self: false }, span: src/builder.rs:55:5: 55:62, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(475) }, name: reserve(520), kind: Method { has_self: true }, span: src/builder.rs:58:5: 58:57, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(487) }, name: add_node(522), kind: Method { has_self: true }, span: src/builder.rs:61:5: 61:42, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(494) }, name: add_nodes(524), kind: Method { has_self: true }, span: src/builder.rs:64:5: 66:6, defaultness: Default { has_value: true } }, TraitItemRef { id: TraitItemId { node_id: NodeId(517) }, name: add_edge(530), kind: Method { has_self: true }, span: src/builder.rs:69:5: 69:72, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(530) }, name: node2id(533), kind: Method { has_self: true }, span: src/builder.rs:72:5: 72:47, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(540) }, name: edge2id(535), kind: Method { has_self: true }, span: src/builder.rs:75:5: 75:47, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(550) }, name: to_graph(537), kind: Method { has_self: true }, span: src/builder.rs:78:5: 78:38, defaultness: Default { has_value: false } }])                                                                                                                   
  --> src/builder.rs:32:1                                                                                                                               
   |                                                                                                                                                    
32 | / pub trait Builder                                                                                                                                
33 | |     where Self: Sized                                                                                                                            
34 | | {                                                                                                                                                
35 | |     /// The graph type produced by this builder.                                                                                                 
...  |
78 | |     fn to_graph(self) -> Self::Graph;
79 | | }
   | |_^

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:450:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
thread 'main' panicked at 'rustc_thread failed: Any', /checkout/src/libcore/result.rs:906:4
error: Could not compile `rs-graph`.

To learn more, run the command again with --verbose.

```

**Comment by oli-obk** [comment]
Can you rerun with `RUST_BACKTRACE=1`?

**Comment by lyro** [comment]
Sure.
```
% env RUST_BACKTRACE=1 cargo clippy
   Compiling rs-graph v0.7.0 (file:///home/rust/rs-graph)
error: internal compiler error: /checkout/src/librustc_typeck/collect.rs:1084: compute_type_of_item: unexpected item type: ItemTrait(Normal, Generics { lifetimes: [], ty_params: [], where_clause: WhereClause { id: NodeId(447), predicates: [BoundPredicate(WhereBoundPredicate { span: src/builder.rs:33:11: 33:22, bound_lifetimes: [], bounded_ty: type(Self), bounds: [TraitTyParamBound(PolyTraitRef { bound_lifetimes: [], trait_ref: TraitRef { path: path(Sized), ref_id: NodeId(449) }, span: src/builder.rs:33:17: 33:22 }, None)] })] }, span: src/lib.rs:1:1: 1:1 }, [], [TraitItemRef { id: TraitItemId { node_id: NodeId(450) }, name: Graph(162), kind: Type, span: src/builder.rs:36:5: 36:16, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(452) }, name: Node(160), kind: Type, span: src/builder.rs:39:5: 39:21, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(455) }, name: Edge(161), kind: Type, span: src/builder.rs:42:5: 42:21, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(458) }, name: new(510), kind: Method { has_self: false }, span: src/builder.rs:45:5: 47:6, defaultness: Default { has_value: true } }, TraitItemRef { id: TraitItemId { node_id: NodeId(466) }, name: with_capacities(511), kind: Method { has_self: false }, span: src/builder.rs:55:5: 55:62, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(475) }, name: reserve(520), kind: Method { has_self: true }, span: src/builder.rs:58:5: 58:57, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(487) }, name: add_node(522), kind: Method { has_self: true }, span: src/builder.rs:61:5: 61:42, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(494) }, name: add_nodes(524), kind: Method { has_self: true }, span: src/builder.rs:64:5: 66:6, defaultness: Default { has_value: true } }, TraitItemRef { id: TraitItemId { node_id: NodeId(517) }, name: add_edge(530), kind: Method { has_self: true }, span: src/builder.rs:69:5: 69:72, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(530) }, name: node2id(533), kind: Method { has_self: true }, span: src/builder.rs:72:5: 72:47, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(540) }, name: edge2id(535), kind: Method { has_self: true }, span: src/builder.rs:75:5: 75:47, defaultness: Default { has_value: false } }, TraitItemRef { id: TraitItemId { node_id: NodeId(550) }, name: to_graph(537), kind: Method { has_self: true }, span: src/builder.rs:78:5: 78:38, defaultness: Default { has_value: false } }])
  --> src/builder.rs:32:1
   |
32 | / pub trait Builder
33 | |     where Self: Sized
34 | | {
35 | |     /// The graph type produced by this builder.
...  |
78 | |     fn to_graph(self) -> Self::Graph;
79 | | }
   | |_^

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:450:8
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
   6: rustc_errors::Handler::span_bug
   7: <std::thread::local::LocalKey<T>>::with
   8: rustc::session::span_bug_fmt
   9: _ZN12rustc_typeck7collect7type_of17hd6c828a23ba3d582E.llvm.6D38641C
  10: rustc::ty::maps::<impl rustc::ty::maps::queries::type_of<'tcx>>::compute_result
  11: _ZN5rustc9dep_graph5graph8DepGraph14with_task_impl17hb0d9810c5b486c1eE.llvm.647F93AF
  12: rustc_errors::Handler::track_diagnostics
  13: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  14: rustc::ty::maps::<impl rustc::ty::maps::queries::type_of<'tcx>>::force
  15: rustc::ty::maps::<impl rustc::ty::maps::queries::type_of<'tcx>>::try_get
  16: rustc::ty::maps::TyCtxtAt::type_of
  17: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::type_of
  18: <clippy_lints::new_without_default::NewWithoutDefault as rustc::lint::LateLintPass<'a, 'tcx>>::check_fn
  19: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_fn
  20: rustc::hir::intravisit::walk_trait_item
  21: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_trait_item
  22: rustc::hir::intravisit::walk_trait_item_ref
  23: rustc::hir::intravisit::walk_item
  24: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  25: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_mod
  26: rustc::hir::intravisit::walk_item
  27: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  28: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_mod
  29: rustc::hir::intravisit::walk_crate
  30: rustc::lint::context::check_crate
  31: rustc::util::common::time
  32: <std::thread::local::LocalKey<T>>::with
  33: <std::thread::local::LocalKey<T>>::with
  34: rustc::ty::context::TyCtxt::create_and_enter
  35: rustc_driver::driver::compile_input
  36: rustc_driver::run_compiler
thread 'main' panicked at 'rustc_thread failed: Any', /checkout/src/libcore/result.rs:906:4
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:69
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:58
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:577
   5: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:538
   6: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:522
   7: rust_begin_unwind
             at /checkout/src/libstd/panicking.rs:498
   8: core::panicking::panic_fmt
             at /checkout/src/libcore/panicking.rs:71
   9: core::result::unwrap_failed
  10: clippy_driver::main
  11: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:99
  12: std::rt::lang_start
             at /checkout/src/libstd/panicking.rs:459
             at /checkout/src/libstd/panic.rs:361
             at /checkout/src/libstd/rt.rs:59
  13: __libc_start_main
  14: _start
error: Could not compile `rs-graph`.

To learn more, run the command again with --verbose.
```

**Comment by oli-obk** [comment]
https://github.com/rust-lang-nursery/rust-clippy/blob/master/clippy_lints/src/new_without_default.rs#L93

should be refactored to use `check_item_impl` for `new` functions, instead of checking for `new` functions and then trying to figure out what impl they belong to.

I'm not sure what we should do if a trait has a `new` method, but probably suggest to add a `: Default` bound on the trait?
